core/tracker/inclusion.go

```diff
@@ -0,0 +1,300 @@
+		}
+	} else if duty.Type == core.DutyAggregator {
+		agg, ok := data.(core.SignedAggregateAndProof)
+		if !ok {
+			return errors.New("invalid aggregate and proof")
+		}
+		attRoot, err = agg.Message.Aggregate.HashTreeRoot()
+		if err != nil {
+			return errors.Wrap(err, "hash aggregate")
+		}
+	}
+
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	i.submissions = append(i.submissions, submission{
+		Duty:    duty,
+		Pubkey:  pubkey,
+		Data:    data,
+		AttRoot: attRoot,
+		Delay:   delay,
+	})
+
+	return nil
+}
+
+// Trim removes all duties that are older than the specified slot.
+// It also calls the missedFunc for any duties that have not been included.
+func (i *inclusion) Trim(ctx context.Context, slot int64) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	var remaining []submission
+	for _, sub := range i.submissions {
+		if sub.Duty.Slot > slot {
+			remaining = append(remaining, sub)
+			continue
+		}
+		if !sub.Included {
+			i.missedFunc(ctx, sub)
+		}
+	}
+	i.submissions = remaining
+}
+
+// CheckBlock checks whether the block includes any of the submitted duties.
+func (i *inclusion) CheckBlock(ctx context.Context, block block) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	for j, sub := range i.submissions {
+		if sub.Included {
+			continue
+		}
+
+		switch sub.Duty.Type {
+		case core.DutyAttester, core.DutyAggregator:
+			_, ok := block.Attestations[sub.AttRoot]
+			if !ok {
+				continue
+			}
+			i.submissions[j].Included = true
+			i.attIncludedFunc(ctx, sub, block)
+		case core.DutyProposer, core.DutyBuilderProposer:
+			if sub.Duty.Slot != block.Slot {
+				continue
+			}
+			i.submissions[j].Included = true
+			// Nothing to report for block inclusions
+		default:
+			panic("bug: unexpected type") // Sanity check, this should never happen
+		}
+	}
+}
+
+// reportMissed reports duties that were broadcast but never included on chain.
+func reportMissed(ctx context.Context, sub submission) {
+	inclusionMisses.WithLabelValues(sub.Duty.Type.String()).Inc()
+
+	switch sub.Duty.Type {
+	case core.DutyAttester, core.DutyAggregator:
+		msg := "Broadcasted attestation never included in any block"
+		if sub.Duty.Type == core.DutyAggregator {
+			msg = "Broadcasted attestation aggregate never included in any block"
+		}
+
+		log.Warn(ctx, msg, nil,
+			z.Any("pubkey", sub.Pubkey),
+			z.I64("attestation_slot", sub.Duty.Slot),
+			z.Any("broadcast_delay", sub.Delay),
+		)
+	case core.DutyProposer, core.DutyBuilderProposer:
+		msg := "Broadcasted block never included in the chain"
+		if sub.Duty.Type == core.DutyBuilderProposer {
+			msg = "Broadcasted blinded block never included in the chain"
+		}
+
+		log.Warn(ctx, msg, nil,
+			z.Any("pubkey", sub.Pubkey),
+			z.I64("block_slot", sub.Duty.Slot),
+			z.Any("broadcast_delay", sub.Delay),
+		)
+	default:
+		panic("bug: unexpected type") // Sanity check, this should never happen
+	}
+}
+
+func reportAttInclusion(ctx context.Context, sub submission, block block) {
```

Comment:
no need to pass full block, just pass `blockSlot`. `AttSlot` can be taken from submission.

Reply:
Also add godoc

Reply:
Does need more stuff